### ISSUE TITLE
Typescript server is slow on Zed compared to VSCode

### ISSUE THREAD

**Issue by baptisteArno** [zed-industries/zed#18698]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

The coding experience in a large Typescript monorepo (https://github.com/baptisteArno/typebot.io) is still an issue for me even though it now uses the vtsls, I still don't get the same speed as VSCode or even NeoVim (with vtsls lsp)

I'm just helpless, what is doing Zed that could cause this code error refresh so slow?

This also can be linked with:
- https://github.com/zed-industries/zed/issues/18621

### Environment

Zed: v0.156.0 (Zed Preview)
OS: macOS 15.0.0
Memory: 16 GiB
Architecture: aarch64

### If applicable, add mockups / screenshots to help explain present your vision of the feature

VSCode:

https://github.com/user-attachments/assets/82b329fc-e36d-4dca-ad89-9b8091198839

Zed:

https://github.com/user-attachments/assets/df83d644-848d-45e6-9da1-ce6ddf824030

**Comment by osiewicz** [comment]
Could you also record a comparison with Neovim + VTSLS? VSC uses tsserver directly.

**Comment by baptisteArno** [comment]
https://github.com/user-attachments/assets/265ac38b-45e9-41c8-90fd-4b8e4ed68321

LazyVim distro with that lsp config override:

```lua
return {
  "neovim/nvim-lspconfig",
  opts = function(_, opts)
    opts.inlay_hints.enabled = false
    opts.servers.vtsls.settings.typescript.tsserver = {
      experimental = {
        enableProjectDiagnostics = false,
      },
    }

    opts.servers.vtsls.autoUseWorkspaceTsdk = true

    return opts
  end,
}
```

**Comment by juanpmarin** [comment]
Same issue here

**Comment by chungweileong94** [comment]
Been facing the same issue for the past few months. The only workaround right now is to restart the LSPs once in a while.

**Comment by chungweileong94** [comment]
I found this [PR from vtsls](https://github.com/yioneko/vtsls/pull/206). It seems Electron came with an unofficial Node version that came with pointer compression, which is also what VSCode is using. I played with it for a while. It does seem like the memory is relatively smaller, but I haven't really started coding with it.

https://github.com/user-attachments/assets/00178251-7b3d-4d2a-996c-e35023840996

**Comment by chungweileong94** [comment]
I'm not sure if this is good enough to reproduce the problem, but if you have a really large monorepo, try opening multiple files from multiple different sub-repos, and you will start noticing that the node memory usage will start to build up, which I think tsserver will likely get slow and unusable like the issue described.

**Comment by wesguirra** [comment]
Also happening for me

**Comment by baptisteArno** [comment]
I wanted to give it another shot. New Zed install, dumb example but it's like this everywhere:

https://github.com/user-attachments/assets/f2c8aa00-a8b2-41ed-9e7b-73024fbb0c0b

Compared with Cursor (or VSCode):

https://github.com/user-attachments/assets/5b218320-9e26-4960-8e15-8b3a5a3bdf61

Guys... How is it not a priority? The reproduction is dead simple. Too bad Zed is so fast but the LSP refresh is total garbage??

@osiewicz, did you manage to reproduce? Clone this https://github.com/baptisteArno/typebot.io (or any non-toy typescript project)

**Comment by baptisteArno** [comment]
Further discussion / debugging here: https://discord.com/channels/869392257814519848/873293828805771284/1343251336778743940

**Comment by aevsai** [comment]
ok, i found what was wrong in my case, zed's default memory limit is different for vtsls and typescript-language-server
you can test if it's works for your case by setting lsp settings to:
"lsp": {
  "vtsls": {
      "settings": {
        "tsserver": {
          "maxTsServerMemory": 8092
        }
      }
    }
}

I'd offer that we set vtsls memory limit same as typescript-language-server
https://github.com/zed-industries/zed/pull/25442

**Comment by chungweileong94** [comment]
> ok, i found what was wrong in my case, zed's default memory limit is different for vtsls and typescript-language-server you can test if it's works for your case by setting lsp settings to:
> 
> ```
> "lsp": {
>   "vtsls": {
>       "settings": {
>         "tsserver": {
>           "maxTsServerMemory": 8092
>         }
>       }
>     }
> }
> ```
> 
> I'd offer that we set vtsls memory limit same as typescript-language-server [#25442](https://github.com/zed-industries/zed/pull/25442)

Interesting, I always though that Zed default the vtsls max memory to 8092, but it doesn't seems to be the case according to your PR🤔
https://zed.dev/docs/languages/typescript#large-projects

**Comment by aevsai** [comment]
then 

> > ok, i found what was wrong in my case, zed's default memory limit is different for vtsls and typescript-language-server you can test if it's works for your case by setting lsp settings to:
> > ```
> > "lsp": {
> >   "vtsls": {
> >       "settings": {
> >         "tsserver": {
> >           "maxTsServerMemory": 8092
> >         }
> >       }
> >     }
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I'd offer that we set vtsls memory limit same as typescript-language-server [#25442](https://github.com/zed-industries/zed/pull/25442)
> 
> Interesting, I always though that Zed default the vtsls max memory to 8092, but it doesn't seems to be the case according to your PR🤔 https://zed.dev/docs/languages/typescript#large-projects
> 
> [EDIT] Actually, I think the max memory is set correctly for vtsls, which is 8092 (https://github.com/zed-industries/zed/pull/17354/files). You probably got confuse by the Zed LSP settings hierarchy.

then i actually not sure what and where went wrong  😵‍💫, bc i tested locally and performance on my repo become great after i edited settings. And that worked in both cases: when it set as defaults in zed and when i edit settings

**Comment by chungweileong94** [comment]
Based on my experience, I don't think memory is the problem here, where one of my monorepos(about 20+ sub-repos) always uses about 5GB at max, which it doesn't even hit the VTSL server max memory. And there is no problem with inspect types, go to definition, etc., not until I start changing some code, where the LSP will slowly fall apart.

**Comment by baptisteArno** [comment]
Memory is not the issue. On nvim with vtsls with default memory limit (4Go), there is no issue.

**Comment by aevsai** [comment]
> Memory is not the issue. On nvim with vtsls with default memory limit (4Go), there is no issue.
maybe my issue was not related from the beginning 😅

**Comment by laneme** [comment]
I believe its related, on my case type changes takes much longer to take effect in other packages in a small monorepo compared to vscode. If the change is in something like generated code (prisma), it takes forever, restarting tsserver/zed works but it becomes a headache very fast.

It doesnt happen in vscode, which could be be because vscode's watchers are more aggressive?
Its a wild guess but my system watcher limit gets maxed out only when vscode is running alongside my programs, not with zed.

**Comment by jpike88** [comment]
maybe it's a good time to consider whether just looking into supporting typescript's upcoming go version? The VSCode extension already supports this

https://github.com/microsoft/typescript-go?tab=readme-ov-file#running-lsp-prototype

**Comment by dohomi** [comment]
I experience the same issue in larger codebase. The autocompletion for component props or component lookup takes more than 5 seconds to show on CTRL+Space

**Comment by PATEL7VISHAL** [comment]
> maybe it's a good time to consider whether just looking into supporting typescript's upcoming go version? The VSCode extension already supports this
> 
> https://github.com/microsoft/typescript-go?tab=readme-ov-file#running-lsp-prototype

I guess, zed should should focus and improve  the LSP stuff, instead of AI bubble, it's clear that zed still not compete with even the **Neovim**.

**Comment by osiewicz** [comment]
We now have an extension for tsgo: https://github.com/zed-extensions/tsgo#

**Comment by Akiyamka** [comment]
I have a fairly large project (about a gigabyte of ts files) and until recently zed was doing great, even faster than vscode. However, the speed of ts degraded over time to the point where it eventually became slower to register for type changes than vscode, and later it became so slow that it became annoying.

Perhaps my problem is a bit different - first of all everything was fine initially, secondly increasing memory for vtsls, as well as switching to tsgo did not give any noticeable changes in speed

**Comment by grushetsky** [comment]
I use Zed on a quite large monorepo project. Unfortunately, I can also experience performance issues with TypeScript code navigation. The major pain point for me is the "Find All References" command. Sometimes it freezes for so long that I need to restart the editor. It works OK after the restart, but then as time goes by it starts to freeze again and again. The autocomplete snappiness doesn't feel great either.

I must admit that Zed is run on a rather outdated MacBook Pro in my case. It's not a vintage one though. And the main thing is that I don't experience these issues in VS Code (which is much more bloated than Zed in my opinion) on the same machine.

I use `vtsls` for TypeScript (and JavaScript). I also tried to adjust the memory dedicated to the TypeScript server (up to 10 GB), but it didn't help. The extra setting that I have configured for TypeScript is `source.fixAll.eslint`. Here's the part of my configuration related to JavaScript and TypeScript:

```json
{
  "languages": {
    "JavaScript": {
      "code_actions_on_format": {
        "source.fixAll.eslint": true
      },
      "language_servers": ["vtsls"]
    },
    "TSX": {
      "code_actions_on_format": {
        "source.fixAll.eslint": true
      }
    },
    "TypeScript": {
      "code_actions_on_format": {
        "source.fixAll.eslint": true
      }
    },
  },
  "lsp": {
    "vtsls": {
      "settings": {
        "javascript": { "tsserver": { "maxTsServerMemory": 10240 } },
        "typescript": { "tsserver": { "maxTsServerMemory": 10240 } }
      }
    }
  }
}
```

It feels to me that the problem has something to do with the binding between the editor and the LSP.

**Comment by thempatel** [comment]
I think the [zed docs](https://zed.dev/docs/languages/typescript#large-projects) are wrong, or rather show settings that don't work.

when i switch from:

```json
{
  "lsp": {
    "vtsls": {
      "settings": {
        // For TypeScript:
        "typescript": { "tsserver": { "maxTsServerMemory": 16184 } },
        // For JavaScript:
        "javascript": { "tsserver": { "maxTsServerMemory": 16184 } }
      }
    }
  }
}
```
to

```json
{
  "lsp": {
    "vtsls": {
      "settings": {
        "tsserver": {
          "maxTsServerMemory": 16184
        }
      }
    }
}
```

it's a complete night and day difference. i haven't dug in, but have to assume that the above settings formulation doesn't actually stick.

**Comment by thempatel** [comment]
> I think the [zed docs](https://zed.dev/docs/languages/typescript#large-projects) are wrong, or rather show settings that don't work.
> 
> when i switch from:
> 
> {
>   "lsp": {
>     "vtsls": {
>       "settings": {
>         // For TypeScript:
>         "typescript": { "tsserver": { "maxTsServerMemory": 16184 } },
>         // For JavaScript:
>         "javascript": { "tsserver": { "maxTsServerMemory": 16184 } }
>       }
>     }
>   }
> }
> to
> 
> {
>   "lsp": {
>     "vtsls": {
>       "settings": {
>         "tsserver": {
>           "maxTsServerMemory": 16184
>         }
>       }
>     }
> }
> it's a complete night and day difference. i haven't dug in, but have to assume that the above settings formulation doesn't actually stick.

whoever is reading this can ignore my previous comment, i think it was likely a fluke. when i look at the `top` output, i can see that max-old-space-size is correctly set to the value defined in settings, and that there are two tsserver processes (1 for `js` and `ts` each).

**Comment by skylovescoffee** [comment]
This vastly improved my speeds:

Update `tsconfig`'s `"include array"` so that it is from the perspective of the root of the project, not the folder it is in.

Example:
```
  "include": [
    "frontend/**/*.ts",
    "frontend/**/*.tsx",
    "frontend/webpack.config.js",
    "frontend/tailwind.config.js",
  ],

```

**Comment by DarkFluid** [comment]
> This vastly improved my speeds:
> 
> Update `tsconfig`'s `"include array"` so that it is from the perspective of the root of the project, not the folder it is in.
> 
> Example:
> 
> ```
>   "include": [
>     "frontend/**/*.ts",
>     "frontend/**/*.tsx",
>     "frontend/webpack.config.js",
>     "frontend/tailwind.config.js",
>   ],
> ```

Tried this for a couple of days. It does not help.
Also my observation is that the LSP's message at the bottom of Zed is quick to print. I think the problem is that the popup, with actions/sugeestions, is waiting for something to complete making it feel slow.

**Comment by qwexs** [comment]
### Solution for TypeScript LSP Performance Issues in Zed

I was experiencing significant TypeScript lag in Zed on my macbook air when working with a large monorepo. The main issue was that when the AI agent modified code, tsserver couldn't keep up with the analysis, leading to memory leaks and slow LSP performance.

**Solution: Using Bun runtime for typescript-language-server**

This dramatically improved performance and significantly reduced RAM usage.

#### Step-by-step instructions:

1. **Install Bun:**
   ```bash
   curl -fsSL https://bun.sh/install | bash
   ```

2. **Get Bun path:**
   ```bash
   which bun
   ```

3. **Install typescript-language-server globally via Bun:**
   ```bash
   bun add -g typescript-language-server
   ```

4. **Configure Zed (settings.json):**
   ```json
   {
     "languages": {
       "TSX": {
         "language_servers": ["typescript-language-server", "!vtsls"],
         "formatter": "prettier",
         "code_actions_on_format": {
           "source.addMissingImports.ts": true,
           "source.organizeImports": false,
           "source.fixAll.eslint": true
         }
       },
       "TypeScript": {
         "language_servers": ["typescript-language-server", "!vtsls"],
         "formatter": "prettier",
         "code_actions_on_format": {
           "source.addMissingImports.ts": true,
           "source.removeUnusedImports": true,
           "source.organizeImports": false,
           "source.fixAll.eslint": true
         }
       }
     },
     "lsp": {
       "typescript-language-server": {
         "binary": {
           "path": "/Users/<USER>/.bun/bin/bun",
           "arguments": [
             "--bun",
             "/Users/<USER>/.bun/bin/typescript-language-server",
             "--stdio"
           ]
         }
       }
     }
   }
   ```

#### Caveats and workarounds:

- **Downside:** Auto-import doesn't appear in the context menu
- **Workaround:** Use `Ctrl + Space` for autocomplete or set `"source.addMissingImports.ts": true` for automatic imports on format

Now Zed performs much better in large monorepositories. I'm looking forward to the final `typescript-go` release, but this solution effectively addresses the performance issues in the meantime.

**Comment by PATEL7VISHAL** [comment]
sorry to say but above solution still seems to be not work for me, feels like default vtsls perform faster then above solution for large project as well, so still zed both typescript LSPs are still slow and feels leggy. Neovim typescript LSPs still faster.

**Comment by chungweileong94** [comment]
> Solution: Using Bun runtime for typescript-language-server
> 
> This dramatically improved performance and significantly reduced RAM usage.

That one of the things that I tried awhile back, unfortunately using bun runtime actually generates more memory usage to me. Probably because of my codebase sizes.

So far the best result that gets is from using NodeJS instance from electron, but tedious to setup.

But there's nothing Zed can do other than waiting tsgo to be released.

**Comment by anthony-c-martin** [comment]
Is it possible that the use of stdin/stdout for IPC is slowing this down? AFAIK VSCode uses named pipes. See https://github.com/zed-industries/zed/blob/5c4f1e6b853013bed14c927aea4a849a8901530e/crates/languages/src/typescript.rs#L589

Looks like the tsgo extension is doing the same thing (tagging @osiewicz):
https://github.com/zed-extensions/tsgo/blob/ef28bd8795a0cc6697013b5a60b32b6aa9201b53/src/tsgo.rs#L166

There's a common convention in the servers I've looked at to support CLI arg `--pipe` (which takes the name of a pipe that the editor opens as a server and waits for a client connection) as an alternative to `--stdio`.

**Comment by ivands** [comment]
TS on zed is becoming unusable for me. It's so slow that it becomes unresponsive.

**Comment by lockieluke** [comment]
Tried `tsgo`, `typescript-language-server` and increasing memory allocation, nothing worked

**Comment by thempatel** [comment]
Huge key piece of info: zed will keep language servers on so long as there is at least 1 file open that is relevant to that language server. I can see why it was built this way, but that means that in a polyglot repo, if you're closing out your `ts` files, the language server is shutdown when the last `ts` file is closed. This means that any cached information the language server keeps around to respond quickly to user requests is destroyed and will have to be totally rebuilt from scratch the next time you open a `ts` file or try to use any LSP feature.

Here's some stuff that's helped move zed from unusable to usable for me:

1. We had some `ts` projects that didn't have an explicit `include` specified. From the [docs](https://www.typescriptlang.org/tsconfig/#include), `include` defaults to `**/*`, which was not what we wanted.
2. Adding `**/node_modules` to my [`file_scan_exclusions`](https://zed.dev/docs/configuring-zed#file-scan-exclusions) (among others) made a huge difference in general responsiveness of the editor.

Between making sure to keep at least 1 typescript file open and 1) and 2) above, `zed` has become fairly usable for TypeScript (for me)!

Here's some resources I found helpful along the way.

1. In case you don't know the full configuration list for `vtsls` (like me), you can find it [here](https://github.com/yioneko/vtsls/blob/cb3c81c7816beb8f24b1ead779d2f3cadf717fc8/packages/service/configuration.schema.json).
2. Enabling `tracing` and setting `log` to `verbose` helped me debug. I noticed that `vtsls` was re-initializing the same `ts` projects over and over again. There were also a bunch of files it was claiming was a part of the project (namely those in `node_modules`). We did have `node_modules` in a global tsconfig exclude, but that didn't seem to prevent this behavior (or we just did it wrong).
3. Here's the [Performance Tracing](https://github.com/microsoft/TypeScript/wiki/Performance-Tracing) guide from the `TypeScript` project. I didn't actually find any performance problems in our code, but if you have some crazy types and your `ts` projects keep getting reinitialized, then you'd be paying that cost over and over again.
4. I added `enableServerSideFuzzyMatch` as an option to mitigate the possibility that having the LSP send data over `stdio` is a bottleneck. Anecdotally, I think it's made a difference but I don't have any hard numbers.

Here's my `vtsls` config in case it's helpful

```json
{
  "lsp": {
    "vtsls": {
      "settings": {
        "vtsls": {
          "autoUseWorkspaceTsdk": true,
          "experimental": {
            "completion": {
              "enableServerSideFuzzyMatch": true,
              "entriesLimit": 10
            }
          }
        },
        "typescript": {
          "preferGoToSourceDefinition": true,
          "tsserver": {
            "maxTsServerMemory": 24576,
            "enableTracing": false,
            "log": "off"
          }
        },
        "javascript": {
          "preferGoToSourceDefinition": true,
          "tsserver": {
            "maxTsServerMemory": 8192,
            "enableTracing": false,
            "log": "off"
          }
        }
      }
    }
  }
}
```

**Comment by DarkFluid** [comment]
> Huge key piece of info: zed will keep language servers on so long as there is at least 1 file open that is relevant to that language server. I can see why it was built this way, but that means that in a polyglot repo, if you're closing out your `ts` files, the language server is shutdown when the last `ts` file is closed. This means that any cached information the language server keeps around to respond quickly to user requests is destroyed and will have to be totally rebuilt from scratch the next time you open a `ts` file or try to use any LSP feature.
> 
> Here's some stuff that's helped move zed from unusable to usable for me:
> 
>     1. We had some `ts` projects that didn't have an explicit `include` specified. From the [docs](https://www.typescriptlang.org/tsconfig/#include), `include` defaults to `**/*`, which was not what we wanted.
> 
>     2. Adding `**/node_modules` to my [`file_scan_exclusions`](https://zed.dev/docs/configuring-zed#file-scan-exclusions) (among others) made a huge difference in general responsiveness of the editor.
> 
> 
> Between making sure to keep at least 1 typescript file open and 1) and 2) above, `zed` has become fairly usable for TypeScript (for me)!
> 
> Here's some resources I found helpful along the way.
> 
>     1. In case you don't know the full configuration list for `vtsls` (like me), you can find it [here](https://github.com/yioneko/vtsls/blob/cb3c81c7816beb8f24b1ead779d2f3cadf717fc8/packages/service/configuration.schema.json).
> 
>     2. Enabling `tracing` and setting `log` to `verbose` helped me debug. I noticed that `vtsls` was re-initializing the same `ts` projects over and over again. There were also a bunch of files it was claiming was a part of the project (namely those in `node_modules`). We did have `node_modules` in a global tsconfig exclude, but that didn't seem to prevent this behavior (or we just did it wrong).
> 
>     3. Here's the [Performance Tracing](https://github.com/microsoft/TypeScript/wiki/Performance-Tracing) guide from the `TypeScript` project. I didn't actually find any performance problems in our code, but if you have some crazy types and your `ts` projects keep getting reinitialized, then you'd be paying that cost over and over again.
> 
>     4. I added `enableServerSideFuzzyMatch` as an option to mitigate the possibility that having the LSP send data over `stdio` is a bottleneck. Anecdotally, I think it's made a difference but I don't have any hard numbers.
> 
> 
> Here's my `vtsls` config in case it's helpful
> 
> {
>   "lsp": {
>     "vtsls": {
>       "settings": {
>         "vtsls": {
>           "autoUseWorkspaceTsdk": true,
>           "experimental": {
>             "completion": {
>               "enableServerSideFuzzyMatch": true,
>               "entriesLimit": 10
>             }
>           }
>         },
>         "typescript": {
>           "preferGoToSourceDefinition": true,
>           "tsserver": {
>             "maxTsServerMemory": 24576,
>             "enableTracing": false,
>             "log": "off"
>           }
>         },
>         "javascript": {
>           "preferGoToSourceDefinition": true,
>           "tsserver": {
>             "maxTsServerMemory": 8192,
>             "enableTracing": false,
>             "log": "off"
>           }
>         }
>       }
>     }
>   }
> }

Unfortunately this hasn't improved anything for me.

**Comment by thempatel** [comment]
@DarkFluid @neyb it would be super helpful if you could

1. provide more details about your setup
2. describe what you've already tried
3. share debug logs from the language server

it will be pretty hard to find a solution without this kind of information!

**Comment by neyb** [comment]
I believe a lot of people are leaving zed because of this ( I presented zed to a friend which stop it because it was too slow, also this Reddit : https://www.reddit.com/r/ZedEditor/s/bOAqxm6kFL).

> provide more details about your setup

Arch Wayland, recent and powerfully AMD Apu. Just open a ts project, auto completion is slow.

My friend had the same problem with another project, on windows with Intel + nvidia

> describe what you've already tried

Increasing ts max memory.

> share debug logs from the language server

I believe the logs are normal. I'll give you more data when I'll be on my computer.

**Comment by tinomarques** [comment]
Hey guys, adding to this conversation as a user who does not want to give up on Zed, but is honestly on the verge of giving up everyday because I just can't get very basic functionality like "Find All References" and member access autocomplete without 10 second waits and constantly pressing "Stop all servers" -> "Restart all servers". It's a joke that the AI autocomplete is faster than the basic typescript autocomplete.

Sorry for how unconstructive this comment is, but my emotions as a user are real and I know this product is turning away heaps of people for getting basic stuff like this so wrong. This thread is nearing 2 years old.

Yes, I've tried every hack in this thread, multiple times.

Here is a short log file @yeskunall with trace on:
[Zed.log](https://github.com/user-attachments/files/26629184/Zed.log)

**Comment by laurensdc** [comment]
I came to Zed because the TypeScript LSP was hanging my Neovim all the time 😓

**Comment by GoldStrikeArch** [comment]
I was curious and decided to take a look on the original setup from the issue which is [typescript monorepo](https://github.com/baptisteArno/typebot.io) and based on the video it is [this function](https://github.com/baptisteArno/typebot.io/blob/main/packages/whatsapp/src/resumeWhatsAppFlow.ts#L511) which I guess was compared with VS code. 

I have Mac M4 Pro with 24 GB RAM and this Zed version (current latest)

```
Zed 0.233.10 
Commit: 350f33871e84d5ed83a403523e9490b9bf014a94
Version: 0.233.10+stable.250.350f33871e84d5ed83a403523e9490b9bf014a94
```

I tried the same "remove parameters, see type errors, add parameters back" exercise and so far I don't see any *meaningful* difference between the VS Code and Zed in terms of "type errors red lines responsiveness"

VS Code:

https://github.com/user-attachments/assets/6049636a-f7c6-42f4-a918-f5962c20c37d

Zed:

https://github.com/user-attachments/assets/8e14c090-b2fb-4bcd-be32-b64e18123bde

cc @baptisteArno

**Comment by svatem1** [comment]
happens to me as well, it renders the product unusable for normal work

**Comment by GoldStrikeArch** [comment]
@svatem1 Can you please try it out on this open source repo (from the original post) -> https://github.com/baptisteArno/typebot.io because I tried it out and so far it is working fine

**Comment by svatem1** [comment]
@GoldStrikeArch It's happenning to me with 1.0.1 Zed on Macos (M1) in my repo consisting of literally couple of lines. Switching to typescript-language-server from vtsls fixes the issue. Updating the display of errors takes maybe 30+ seconds with vtsls.

**Comment by GoldStrikeArch** [comment]
Don't know what to say, there should be some reproducible example otherwise it is almost impossible to say what can be wrong.

**Comment by svatem1** [comment]
I'd really like to help, but I have no idea how to make the problem reproducible on your side. I just wanted to compile some excerpt from my codebase to demonstrate that issue on, restarted editor while doing so and I'm not able to reproduce it even on the original malfunctioning code. Meaning the underlines are now displayed in sub 5s time consistently. So it seems to me like it's not just specific code but also something that accumulates over time that no matter what I won't be able to provide to you.

**Comment by GoldStrikeArch** [comment]
> I'd really like to help, but I have no idea how to make the problem reproducible on your side

Maybe there are some factors from your end? Like (I am just trying to think about something, so I am making up the stuff) you are using the Zed editor for a whole week without closing it? Or maybe you have 100+ TS files open simultaneously? There should be something because it looks like on the "default regular usage" it should not happen

**Comment by svatem1** [comment]
@GoldStrikeArch Yes, I am using Zed for prolonged period of time, but isn't it common use case? Do you turn off your laptop each time you want to pause working on it? I'm worried what your answer will be. Files in project +-20, opened files: 5, it's a small prototype project with few LOC.

**Comment by GoldStrikeArch** [comment]
>  but isn't it common use case

I am usually open/close it a lot during the day 😅 maybe 10+ times maybe more

> Do you turn off your laptop each time you want to pause working on it

Not sure that lock screen can impact TypeScript LSP, but it is good to know all "specific step by step scenarios" in which that performance downgrade can happen, e.g:
- in 9 AM you open your laptop and Zed was there already (and you didn't re-open it).
- then you worked with small project ~20 opened files. 
- at some point (let's say 5 PM) you started to notice the "unusable LSP performance" or something like that

Because without understanding the whole "how you use it" it is super hard to pinpoint where to look at or what to do

**Comment by PATEL7VISHAL** [comment]
Tested on Zed Stable 1.0.1, and I still notice noticeable lag in TypeScript LSP completions. Compared side-by-side, VS Code still performs best, and even Neovim feels more responsive for large TypeScript projects.

This seems like a general performance issue rather than a project-specific setup problem. I don’t think detailed reproduction steps are necessary here because the difference is easy to observe by comparing Zed against VS Code or Neovim on the same TypeScript workspace.

**Comment by GoldStrikeArch** [comment]
> This seems like a general performance issue rather than a project-specific setup problem. I don’t think detailed reproduction steps are necessary here because the difference is easy to observe by comparing Zed against VS Code or Neovim on the same TypeScript workspace.


Can you please try it out on that specific repo (https://github.com/baptisteArno/typebot.io) from the original post? The thing is that it is suppose to be "laggy" on large monorepos or something like that. I tried that particular repo myself and didn't notice anything, but maybe you will be able to reproduce it easily (which would help a lot)

**Comment by svatem1** [comment]
@GoldStrikeArch maybe the scope/context of the task should shift from "it's slow on this mono/repo because of a specific code it contains" to "it gets significantly slower after prolonged usage"
I don't have monorepo, I have extremely small codebase, yet I was experiencing 30+ seconds lag on M1 Max CPU with 64GB of memory until I restarted the editor. Also - you already were unable to reproduce the problem on that repo yourself, while ppl are still reporting it's happening, so maybe the method you're trying to employ to reproduce the rpoblem is not of help here.

**Comment by GoldStrikeArch** [comment]
@svatem1 

> so maybe the method you're trying to employ to reproduce the rpoblem is not of help here.

The thing is that right now the problem sounds like this: "in some cases (probably after prolonged usage, which is like 5+ hours? 10+ hours?), on some projects, when doing something, the TypeScript LSP is laggy". I am trying to narrow down exactly what I (or anybody else) should do to reproduce/experience at least some portion of it. Because 30+ seconds of lag sounds crazy, maybe there is something like "we don't free the memory of not used files and keep them there forever" and after a long period you just out of memory for vtls, maybe there is some bug in the integration of the LSP itself, and the problem is not only to TypeScript, etc.

So, without some more detailed info it is super hard to approach this problem. I want to fix it but I don't know how I can even experience it, do you see my point here? I tried out a "big TS monorepo", so maybe I need to do it for 5 hours? I don't know because right now the description is super vague and this "it happens after some time" is the only clue you provided.

**Comment by GoldStrikeArch** [comment]
@svatem1 if you experience it constantly after a long period of time, can you please try out a higher value for the memory of vtls? Something like this:

```json
"lsp": {
  "vtsls": {
    "initialization_options": {
      "maxTsServerMemory": 16384 // OR 32768 which is like 16 GB or 32 GB
    }
  }
}
```

It will at least narrow down some potential root causes

**Comment by PATEL7VISHAL** [comment]
@GoldStrikeArch Yes, I have a somewhat larger project: around 400 files, with ~230 TypeScript files and about 58k LOC total (~34k TypeScript LOC).

But when testing the same project in Visual Studio Code or Neovim, everything works smoothly without noticeable lag.

I also tested with the reference repo https://github.com/baptisteArno/typebot.io)
. In that case, I noticed some lag in both Zed and Neovim, which seems expected since the project is significantly larger than mine.

Still, if Zed only performs well in certain environments while other editors consistently provide good performance across environments, I would personally prefer those editors instead of adjusting the project setup specifically to make Zed run smoothly.

I would appreciate it if this is treated as a real bug so progress can be made, rather than focusing only on validating and reproducing the issue.

One thing I noticed is that Zed seems to be improving its LSP behavior.

I noticed that Zed tries to provide completions faster by first showing a quick completion list, then filling in more detailed information later (for example whether something is a method, variable, etc.). It feels like a two-pass completion process.

**Comment by thomasmol** [comment]
> I would appreciate it if this is treated as a real bug so progress can be made, rather than focusing only on validating and reproducing the issue.

@PATEL7VISHAL What do you think the first step to fix a bug is?

**Comment by GoldStrikeArch** [comment]
@PATEL7VISHAL 

> I noticed some lag in both Zed and Neovim

Can you please record it? Because I tried it out and for me it was "almost the same" between VS code and Zed on that repo.

**Comment by svatem1** [comment]
> [@svatem1](https://github.com/svatem1) if you experience it constantly after a long period of time, can you please try out a higher value for the memory of vtls? Something like this:
> 
> "lsp": {
>   "vtsls": {
>     "initialization_options": {
>       "maxTsServerMemory": 16384 // OR 32768 which is like 16 GB or 32 GB
>     }
>   }
> }
> It will at least narrow down some potential root causes

When/if it starts happening again, I'll definitely try this, thank you!

**Comment by svatem1** [comment]
> [@svatem1](https://github.com/svatem1)
> 
> > so maybe the method you're trying to employ to reproduce the rpoblem is not of help here.
> 
> The thing is that right now the problem sounds like this: "in some cases (probably after prolonged usage, which is like 5+ hours? 10+ hours?), on some projects, when doing something, the TypeScript LSP is laggy". I am trying to narrow down exactly what I (or anybody else) should do to reproduce/experience at least some portion of it. Because 30+ seconds of lag sounds crazy, maybe there is something like "we don't free the memory of not used files and keep them there forever" and after a long period you just out of memory for vtls, maybe there is some bug in the integration of the LSP itself, and the problem is not only to TypeScript, etc.
> 
> So, without some more detailed info it is super hard to approach this problem. I want to fix it but I don't know how I can even experience it, do you see my point here? I tried out a "big TS monorepo", so maybe I need to do it for 5 hours? I don't know because right now the description is super vague and this "it happens after some time" is the only clue you provided.

Well I'd start by trying to reproduce the problem the way users (me here eg.) are actually using the editor = not 5 minutes clicking but having it basically run indefinitely (I do not quit unless there's update eg.), with laptop lid being closed couple of times a day (I presume this puts the macbook into some kind of a sleep state). In this particular case, I'd open the repo you linked (some users already had problem with it so it's a good start), do some even small edits, keep the editor running few hours, again do some edits, this cycle couple of times on a span of couple of days, not much action needed and maybe something will show up.

**Comment by laurensdc** [comment]
Issue for me was an import of a barrel file that used `typed-inject` (DI framework) to resolve a bunch of nested dependencies. When we import this barrel file, the LSP goes out of control. Happens in other editors too, but seems worse in Zed. I can't share the code, but would it be worth trying to re-create a similar setup?

**Comment by GoldStrikeArch** [comment]
@laurensdc 

>  can't share the code, but would it be worth trying to re-create a similar setup?

Sure, if you can create a reproducible example it would help a lot

**Comment by hmasovic** [comment]
I just came back to see if these issues were fixed or not, I dropped zed back to vscode/cursor almost 1y ago because of these issues. And also zed was hogging up the cpu/ram causing the M2 laptop to heat up compared to vscode/cursor (you only feel it once you actually start changing the code, moving through files + it's slow on response time - as pointed above). If you're doing little to no work, it's not noticable.

Last time I checked vscode used a more optimized LSP compared to zed. It seems like zed is good for rust projects, not so much typescript projects. Haven't tested it out on Go projects/other languages.

**Comment by laurensdc** [comment]
> Sure, if you can create a reproducible example it would help a lot

I gave it a go but failed to create the same LSP sluggishness that I experience day-to-day in my current client project.
I tried to create the same level of complexity, but LSP seemed to behave just fine.

My gut feeling is that it's a combination of `typed-inject` resolving nested dependency chains and type information + a 200k+ lines Prisma types file, both doing a bunch of nested/generic type magic blowing up the LSP.

**Comment by GoldStrikeArch** [comment]
@laurensdc 
> My gut feeling is that it's a combination of `typed-inject` resolving nested dependency chains and type information + a 200k+ lines Prisma types file, both doing a bunch of nested/generic type magic blowing up the LSP.

This sounds like it should lag with other IDE as well. Maybe you can try to use the [tsgo](https://github.com/microsoft/typescript-go) and see if the new LSP will be faster for you?

**Comment by laurensdc** [comment]
> This sounds like it should lag with other IDE as well. Maybe you can try to use the [tsgo](https://github.com/microsoft/typescript-go) and see if the new LSP will be faster for you?

When I tried that it spawned a memory hogging process that kept eating more and more RAM even after closing Zed, to the point where it froze my machine.. So I went back to defaults as much as possible. Maybe it's our project, maybe it's just TS, I don't know..

**Comment by kayakyakr** [comment]
Having to abandon Zed due to this. I rolled over to tsgo to see how it worked. It is certainly _much_ more performant than vtsls was, and it's only too happy to run its RAM usage up to 16 gigs, but it still is _rough_ on actually coding. It idles at 6.5g. 

I've actually got Zed itself idling at 13 gigs of RAM with my project open as well. Launching a type suggestion popup and using the arrow keys to navigate got the entire UI to freeze and roll up to 24 gigs of memory.

Still will use Zed for non-monorepo, smaller projects, but for the big one? I guess I'm going back to vscode.

**Comment by GoldStrikeArch** [comment]
@kayakyakr 
> I've actually got Zed itself idling at 13 gigs of RAM with my project open as well. Launching a type suggestion popup and using the arrow keys to navigate got the entire UI to freeze and roll up to 24 gigs of memory.

> Still will use Zed for non-monorepo, smaller projects, but for the big one? I guess I'm going back to vscode.

Can you please check this monorepo from the original post -> https://github.com/baptisteArno/typebot.io. Do you have a performance issues there with TypeScript LSP (default vtls)?

**Comment by avirankatz** [comment]
Why not switching from vtsls to typescript-language-server as the default TS LSP?

**Comment by saverioscagnoli** [comment]
I fixed it by installing the [tsgo](https://github.com/zed-extensions/tsgo) extension and putting this in the settings file:
```json
    "TypeScript": {
      "language_servers": [
        "typescript-ls",
        "!vtsls",
        "!typescript-language-server",
      ],
    },
    "TSX": {
      "language_servers": [
        "typescript-ls",
        "!vtsls",
        "!typescript-language-server",
      ],
    },
```
